pkg/credentialprovider/plugin/examples/acr/main.go

```diff
@@ -0,0 +1,99 @@
+// +build !providerless
+
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// TODO: this is just a reference implementation, this plugin should be moved
```

Comment:
@feiskyer this is a reference implementation for azure container registry just to validate whether the proposed plugin mechanism is feasible and can support existing features in the Azure credential provider. It should be moved to kubernetes-sigs/cloud-provider-azure at some point.

Can you review this and provide feedback? In particular I'm not sure what the expected behavior around anonymous authentication is.

Reply:
I think current implementation is OK. Once we have the plugin available for e2e, we could validate more in the e2e tests.

Reply:
For anonymous authn, we need to append a dummy DockerConfigEntry, e.g. `cfg["*.azurecr.*"] = defaultConfigEntry`

Reply:
@andrewsykim could the API support a list of responses in the criauth API?

Reply:
> For anonymous authn, we need to append a dummy DockerConfigEntry, e.g. cfg["*.azurecr.*"] = defaultConfigEntry

Is anonymous auth still needed if valid credentials are returned from managed identity? 

> @andrewsykim could we support a list of responses in the criauth API?

I can look into this, but I avoided it initially because it complicates the way the kubelet caches mulitple entries. Should it cache per entry or all credentials returned from the plugin?

Reply:
anonymous auth is still required, see the comments here https://github.com/kubernetes/kubernetes/pull/94196#discussion_r496633710

Reply:
That comment is asking to add it back but it doesn't explain why it's required. 

I'm wondering whether anonymous auth is required if valid credentials are also returned. In other words, is it okay to return anonymous auth only as a fallback or do we always have to return both? It sounds like we have to return both but I would like to understand better why.

Reply:
@andrewsykim anonymous auth is still required, that's a requirement from Azure Container Registry(ACR). If ACR only allows anonymous auth and invalid credentials(from k8s side, we don't know whether one ACR only allows anonymous auth or not) would make pull image failure.

Reply:
Thanks for clarifying, I'll try to update the API so multiple AuthConfigs can be returned.